### DISCUSSION TITLE
Deploy NextJS on CPanel

### DISCUSSION THREAD

**Discussion by DanielPenalozaB** [vercel/next.js#55411] [Help] [answered]
### Summary

Hello everyone,

I'm trying to upload a basic NextJS application to a CPanel just to test the functionalities of this framework and hopelessly start uploading real projects.

I have been following the NextJS/Vercel docs searching for a guide to deploy my app to a CPanel self hosted VPS but don't know where to start and what configurations to follow. Does anyone knows or has a guide that helps to achive that?

My CPanel does not have a Node.js software to manage the SPA or Web Applications so it's likely by terminal. This is the basic configuration of my project:

# Initialization

This is the initial configuration of the landing page project that I setted.

```
√ Would you like to use TypeScript? ... No
√ Would you like to use ESLint? ... No
√ Would you like to use Tailwind CSS? ... Yes
√ Would you like to use `src/` directory? ... Yes
√ Would you like to use App Router? (recommended) ... Yes
? Would you like to customize the default import alias? » No
```

# next.config.js

> I don't get any toruble while doing the export, all works good.

Selected the `"output": "export"` to get the files ready to upload to CPanel (like I used to do it with `npm run build` in React Vanilla).

```
/** @type {import('next').NextConfig} */
const nextConfig = {
	output: "export",
};

module.exports = nextConfig;
```

# Folder structure

```
node_modules
public
src /
    app/
    about/
        page.js
    favicon.ico
    global.css
    layout.js
    page.js
```

# .htaccess

```
RewriteBase /
RewriteRule ^index\.html$ - [L]
 RewriteCond %{REQUEST_FILENAME} !-f
 RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule . /index.html [L]
```

# Troubles

When I upload the exported project in the folder "out/" and upload it to the CPanel with the custom .htaccess (like i used to do with React Vanilla) the index loads great, but when I try to go to other view by URL (like: "https://subdomain.domain.com/about") or when I go to other view (using the useRouter() hook) and reload the page it renders the index content while keeps the `"/about"` URL.

The only way to render the view data is using the `router.push("/about")`, otherwise it shows the index content.

How can I avoid this and get a similar behavior like using React vanilla?

# Good to know

The CPanel is self hosted, it's a basic plan so we don't have the Node.js software to manage the web applications and the hosting provider says that they can't take part in our self hosted VPS (shamesly).

I installed Node.js through the terminal so the server has the 16 version.

There is no special functionalities in the code, just the `router.push()` from  `router/navigation` hook, plain HTML rendered with tailwind styles and that is.

If there is any other information needed feel free to ask, thanks a lot.

**Comment by icyJoseph** [comment]
The difference between this and what you call "vanilla" is that export is producing an entry point for every page you have. 

It is often the case that that you call "vanilla" React apps, only have one entry point. And the Router takes over client side, and it decides what to show then. 

You need to make it so that your .htaccess or CPanel settings do not redirect every thing to `/`. You just need to map something like `/about.html` to `/about`, otherwise reloading the page will also fail.

I'd say, first try without the .htaccess file, what do you see then? and then fix that instead.

**Reply by DanielPenalozaB** [reply]
It worked! Thanks for the approach.

Now I have this doubt, I'm doing a landing page with lang/locale functionality, that switches between English and Spanish language.

## Project structure

I'm using the [NextJS localization](https://nextjs.org/docs/app/building-your-application/routing/internationalization#localization) routing structure, so I'm using`server` and `client` functionalities to achieve this, and in local works perfect, but when I upload it to web (the same way as I metioned before) I can't get the routes, I assume it's the same reason but don't really know how to fix this behavior.

### Folder structure

The [lang] directory stands for `/en` & `/es` language routes.

![image](https://github.com/vercel/next.js/assets/118763939/176f54ff-9620-4ed9-ba46-a5ced2792904)

### i18n.config.js

```
export const i18n = {
    defaultLocale: "es",
    locales: ["es", "en"]
}
```

### getDictonary.js

```
import 'server-only';

const dictionaries = {
    es: () => import('../dictionaries/es.json').then((module) => module.default),
    en: () => import('../dictionaries/en.json').then((module) => module.default),
};

export const getDictionary = async (locale) => dictionaries[locale]();
```

### dictionaries

![image](https://github.com/vercel/next.js/assets/118763939/5c9c8644-c3c5-4d25-b612-24008f00dcc7)


## Out folder

This is the build directories and files result.

![image](https://github.com/vercel/next.js/assets/118763939/30ffa568-322b-42c4-8f3d-9d946f414ab8)

![image](https://github.com/vercel/next.js/assets/118763939/54e8a670-26b2-4d12-af4a-8f3a18a74af2)

## Troubles

Now that I don't have a static files structure, but a complex one, how could I config the .htaccess in this case? Pointing to achieve a good behavior for the web.

**Reply by icyJoseph** [reply]
I'm out-of-office right now, but that sounds like a server feature? If you try locally to build and then use this tool for the .out folder contents:

- https://www.npmjs.com/package/serve

It won't work either right? So perhaps that feature you describe is not supported for export mode? I'm not 100% sure.

**Reply by DanielPenalozaB** [reply]
Your right, It doesn't work, I only get this messages. I'll research more about this functionality and it's scope but please let me know if you know something about it.

![image](https://github.com/vercel/next.js/assets/118763939/a30f6996-8329-4371-8bca-2468a3c16f93)
![image](https://github.com/vercel/next.js/assets/118763939/34183971-fc45-44a7-bd16-c218c222b970)

Thanks for the approach.

**Reply by icyJoseph** [reply]
Sure thing. Though if your main concern is solved please don't forget to mark this as answered.

Feel free to keep on asking :)

**Reply by DanielPenalozaB** [reply]
Thanks a lot, I learned so much!

**Reply by valentindlr1** [reply]
Daniel, I'm dealing with the exact same problem right now. Were you able to solve this?

**Reply by DanielPenalozaB** [reply, marked as answer]
Hello @valentindlr1, more or less.

# Why is this happening?

The problem is that NextJs builds contain both client and server components, and the only way to run a NextJs project (involving server components or dynamic routing) in a production environment is to configure a server that listens for requests and drive, that's where we get stuck.

There are certain ways to avoid or overcome this situation, some of them (that I know of) are the following:

## 1. Deploy using Vercel CLI.

This is a really easy way to deploy a NextJs project to production, the interface and process is really friendly and intuitive and you can point it to its own domain.

### How to implement using Vercel?

[Deploy your Nextjs using Vercel CLI](https://www.youtube.com/watch?v=W3jKJ3V_4V4&t=2s)

>By the way, this is a great YouTube channel, subscribe to Sonny; )

[Vercel CLI Deployment Documentation](https://nextjs.org/docs/app/building-your-application/deploying#managed-nextjs-with-vercel)

## 2. Deploy using Nginx with pm2.

This is a more complex option, but provides more "control" over your project in production, the basic logic behind this is to upload all your project code to your hosting services (VPS in my case) and run `npm run start ` (or your package manager script) and now with your project available on the web you use `pm2` to keep your project running.

### How to deploy to a custom server?

[How to deploy a Next.js app to a custom server – NOT Vercel! (Complete tutorial for beginners)](https://www.youtube.com/watch?v=HIb4Ucs_foQ)

>Again Sonny, amazing content.

[https://www.youtube.com/watch?v=kHL1FGg0XDA](https://www.youtube.com/watch?v=kHL1FGg0XDA)

[https://www.youtube.com/watch?v=kHL1FGg0XDA](https://www.youtube.com/watch?v=DbREUJNgxqw&t=469s)


# My experience

I get a VPS service through hostinger and get my Nextjs projects to work correctly (I used te custom server guides). At first it's challenging but it's a really logic and easy process.

I hope it helps.

Anyway, always read the official Vercel Nextjs documentation. [docs](https://nextjs.org/docs/app/building-your-application/deploying)
